File: android/walletlink/src/main/java/com/coinbase/walletlink/apis/WalletLinkWebSocket.kt
package com.coinbase.walletlink.apis

import com.coinbase.wallet.core.extensions.asJsonMap
import com.coinbase.wallet.core.extensions.logError
import com.coinbase.wallet.core.extensions.retryWithDelay
import com.coinbase.wallet.core.extensions.takeSingle
import com.coinbase.wallet.core.util.ConcurrentLruCache
import com.coinbase.wallet.http.connectivity.Internet
import com.coinbase.wallet.http.models.WebIncomingDataType
import com.coinbase.wallet.http.models.WebIncomingText
import com.coinbase.wallet.http.websocket.WebSocket
import com.coinbase.walletlink.dtos.ClientResponseDTO
import com.coinbase.walletlink.dtos.JoinSessionMessageDTO
import com.coinbase.walletlink.dtos.PublishEventDTO
import com.coinbase.walletlink.dtos.ServerRequestDTO
import com.coinbase.walletlink.dtos.SetMetadataMessageDTO
import com.coinbase.walletlink.dtos.SetSessionConfigMessageDTO
import com.coinbase.wallet.core.interfaces.JsonSerializable
import com.coinbase.walletlink.models.ClientMetadataKey
import com.coinbase.walletlink.models.EventType
import com.coinbase.walletlink.models.ServerMessageType
import io.reactivex.Single
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.rxkotlin.addTo
import io.reactivex.schedulers.Schedulers
import io.reactivex.subjects.PublishSubject
import io.reactivex.subjects.ReplaySubject
import java.net.URL
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicInteger

private data class WalletLinkCallback(val requestId: Int, val subject: ReplaySubject<ClientResponseDTO>)

/**
 * Represents a WalletLink WebSocket connection
 */
internal class WalletLinkWebSocket(val url: URL) {
    private val serialScheduler = Schedulers.single()
    private val disposeBag = CompositeDisposable()
    private val connection = WebSocket(url)
    private var callbackSequence = AtomicInteger()
    private var incomingRequestsSubject = PublishSubject.create<ServerRequestDTO>()
    private var pendingCallbacks = ConcurrentLruCache<Int, ReplaySubject<ClientResponseDTO>>(maxSize = 300)

    /**
     * Incoming WalletLink requests
     */
    val incomingRequestsObservable = incomingRequestsSubject.hide()

    /**
     * WalletLink Connection state
     */
    val connectionStateObservable = connection.connectionStateObservable

    /**
     * Connect to WalletLink server
     */
    fun connect(): Single<Unit> {
        disposeBag.clear()

        connection.incomingObservable
            .observeOn(serialScheduler)
            .subscribe { processIncomingData(it) }
            .addTo(disposeBag)

        return connection.connect()
    }

    /**
     * Disconnect from WalletLink server
     */
    fun disconnect(): Single<Unit> {
        disposeBag.clear()
        return connection.disconnect()
    }

    /**
     * Join a WalletLink session
     *
     * @param sessionKey sha256(session+secret) hash
     * @param sessionId session ID scanned offline (QR code, NFC, etc)
     *
     * @return A single wrapping `Boolean` to indicate operation was successful
     */
    fun joinSession(sessionKey: String, sessionId: String): Single<Boolean> {
        val callback = createCallback()
        val message = JoinSessionMessageDTO(id = callback.requestId, sessionId = sessionId, sessionKey = sessionKey)

        return send(message, callback = callback)
    }
    /**
    * Set metadata in the current session
    *
    *  @param key Metadata key on WalletLink server
    *  @param value Metadata value stored on WalletLink server. This data may be encrypted.
    *  @param sessionId Session ID scanned offline (QR code, NFC, etc)
    *
     * @return A single wrapping `Boolean` to indicate operation was successful
     */
    fun setMetadata(key: ClientMetadataKey, value: String, sessionId: String): Single<Boolean> {
        val callback = createCallback()
        val id = callback.requestId
        val message = SetMetadataMessageDTO(id = id, sessionId = sessionId, key = key.rawValue, value = value)

        return send(message, callback = callback)
    }

    /**
     * Set session config once a link is established
     *
     * @param webhookId Webhook ID used to push notifications to mobile client
     * @param webhookUrl Webhook URL used to push notifications to mobile client
     * @param metadata Metadata forwarded to host
     * @param sessionId Session ID scanned offline (QR code, NFC, etc)
     *
     * @return A single wrapping `Boolean` to indicate operation was successful
     */
    fun setSessionConfig(
        webhookId: String,
        webhookUrl: URL,
        metadata: Map<String, String>,
        sessionId: String
    ): Single<Boolean> {
        val callback = createCallback()
        val message = SetSessionConfigMessageDTO(
            id = callback.requestId,
            sessionId = sessionId,
            webhookId = webhookId,
            webhookUrl = webhookUrl.toString(),
            metadata = metadata
        )

        return send(message, callback = callback)
    }

    /**
     * Publish new event to WalletLink server
     *
     * @param event Event type to publish
     * @param data The encrypted data sent to host
     * @param sessionId Session ID scanned offline (QR code, NFC, etc)
     *
     * @return A single wrapping `Boolean` to indicate operation was successful
     */
    fun publishEvent(event: EventType, data: String, sessionId: String): Single<Boolean> {
        val callback = createCallback()
        val message = PublishEventDTO(id = callback.requestId, sessionId = sessionId, event = event, data = data)

        return send(message, callback = callback)
    }

    // Send message helper(s)

    private fun send(message: JsonSerializable, callback: WalletLinkCallback): Single<Boolean> {
        val jsonString = message.asJsonString()

        return Internet.statusChanges
            .filter { it.isOnline }
            .takeSingle()
            .flatMap { connection.sendString(jsonString) }
            .flatMap { callback.subject.takeSingle() }
            .map { it.type.isOK }
            .retryWithDelay(3, 1, TimeUnit.SECONDS)
            .timeout(15, TimeUnit.SECONDS)
            .logError()
            .map { success ->
                pendingCallbacks.remove(callback.requestId)
                success
            }
            .onErrorResumeNext { err ->
                pendingCallbacks.remove(callback.requestId)
                throw err
            }
    }

    // Pending requests callback management

    private fun createCallback(): WalletLinkCallback {
        val requestId = callbackSequence.incrementAndGet()
        val subject = ReplaySubject.create<ClientResponseDTO>(1)

        pendingCallbacks[requestId] = subject

        return WalletLinkCallback(requestId = requestId, subject = subject)
    }

    /**
     * This is called when the host sends a response to a request initiated by the signer
     */
    private fun receivedClientResponse(response: ClientResponseDTO) {
        val requestId = response.id ?: return
        val subject = pendingCallbacks[requestId]

        pendingCallbacks.remove(requestId)

        subject?.onNext(response)
    }

    /**
     * this is called when the host sends a request initiated by the host
     */
    private fun receivedServerRequest(request: ServerRequestDTO) {
        incomingRequestsSubject.onNext(request)
    }

    private fun processIncomingData(incoming: WebIncomingDataType) {
        val incomingText = incoming as? WebIncomingText ?: return
        val jsonString = incomingText.string
        val json: Map<String, Any> = jsonString.asJsonMap() ?: return
        val typeString: String = json["type"] as? String ?: return
        val messageType: ServerMessageType = ServerMessageType.fromRawValue(typeString) ?: return

        when (messageType) {
            ServerMessageType.OK, ServerMessageType.Fail, ServerMessageType.PublishEventOK -> {
                val response = ClientResponseDTO.fromJsonString(jsonString) ?: return
                receivedClientResponse(response)
            }
            ServerMessageType.Event -> {
                val request = ServerRequestDTO.fromJsonString(jsonString) ?: return
                receivedServerRequest(request)
            }
        }
    }
}

File: ios/Source/APIs/WalletLinkWebSocket.swift
// Copyright (c) 2017-2019 Coinbase Inc. See LICENSE

import CBCore
import CBHTTP
import RxSwift

private typealias WalletLinkCallback = (requestId: Int32, subject: ReplaySubject<ClientResponseDTO>)

/// Represents a WalletLink WebSocket
final class WalletLinkWebSocket {
    private let serialScheduler = SerialDispatchQueueScheduler(internalSerialQueueName: "WalletLinkWebSocket.serial")
    private let connection: WebSocket
    private var disposeBag = DisposeBag()
    private var callbackSequence = AtomicInt32()
    private var incomingRequestsSubject = PublishSubject<ServerRequestDTO>()
    private var pendingCallbacks = BoundedCache<Int32, ReplaySubject<ClientResponseDTO>>(maxSize: 300)

    /// Incoming WalletLink requests
    let incomingRequestsObservable: Observable<ServerRequestDTO>

    /// WalletLink Connection state
    let connectionStateObservable: Observable<WebConnectionState>

    /// Constructor
    ///
    /// - Parameters:
    ///     - url: WalletLink WebSocket URL
    required init(url: URL) {
        let connection = WebSocket(url: url)

        self.connection = connection
        incomingRequestsObservable = incomingRequestsSubject.asObservable()
        connectionStateObservable = connection.connectionStateObservable
    }

    /// Connect to WalletLink server
    func connect() -> Single<Void> {
        disposeBag = DisposeBag()

        connection.incomingObservable
            .observeOn(serialScheduler)
            .subscribe(onNext: { [weak self] in self?.processIncomingData($0) })
            .disposed(by: disposeBag)

        return connection.connect()
    }

    /// Disconnect from WalletLink server
    func disconnect() -> Single<Void> {
        disposeBag = DisposeBag()
        return connection.disconnect()
    }

    /// Join a WalletLink session
    ///
    /// - Parameters:
    ///   - sessionKey: sha256(session+secret) hash
    ///   - sessionId: session ID scanned offline (QR code, NFC, etc)
    ///
    /// - Returns: A single wrapping `Boolean` to indicate operation was successful
    func joinSession(using sessionKey: String, for sessionId: String) -> Single<Bool> {
        let callback = createCallback()
        let message = JoinSessionMessageDTO(id: callback.requestId, sessionId: sessionId, sessionKey: sessionKey)

        return send(message, callback: callback)
    }

    /// Set metadata in the current session
    ///
    /// - Parameters:
    ///   - key: Metadata key on WalletLink server
    ///   - value: Metadata value stored on WalletLink server. This is encrypted.
    ///   - sessionId: Session ID scanned offline (QR code, NFC, etc)
    ///
    /// - Returns: A single wrapping `Boolean` to indicate operation was successful
    func setMetadata(key: ClientMetadataKey, value: String, for sessionId: String) -> Single<Bool> {
        let callback = createCallback()
        let key = key.rawValue
        let message = SetMetadataMessageDTO(id: callback.requestId, sessionId: sessionId, key: key, value: value)

        return send(message, callback: callback)
    }

    /// Set session config once a link is established
    ///
    /// - Parameters:
    ///   - webhookId: Webhook ID used to push notifications to mobile client
    ///   - webhookUrl: Webhook URL used to push notifications to mobile client
    ///   - metadata: Metadata forwarded to host
    ///   - sessionId: Session ID scanned offline (QR code, NFC, etc)
    ///
    /// - Returns: A single wrapping `Boolean` to indicate operation was successful
    func setSessionConfig(
        webhookId: String,
        webhookUrl: URL,
        metadata: [String: String],
        for sessionId: String
    ) -> Single<Bool> {
        let callback = createCallback()
        let message = SetSessionConfigMessageDTO(
            id: callback.requestId,
            sessionId: sessionId,
            webhookId: webhookId,
            webhookUrl: webhookUrl.absoluteString,
            metadata: metadata
        )

        return send(message, callback: callback)
    }

    /// Publish new event to WalletLink server
    ///
    /// - Parameters:
    ///   - event: Event type to publish
    ///   - data: The encrypted data sent to host
    ///   - sessionId: Session ID scanned offline (QR code, NFC, etc)
    ///
    /// - Returns: A single wrapping `Boolean` to indicate operation was successful
    func publishEvent(_ event: EventType, data: String, to sessionId: String) -> Single<Bool> {
        let callback = createCallback()
        let message = PublishEventDTO(id: callback.requestId, sessionId: sessionId, event: event, data: data)

        return send(message, callback: callback)
    }

    // MARK: - Send message helper(s)

    private func send(_ message: JSONSerializable, callback: WalletLinkCallback) -> Single<Bool> {
        guard let jsonString = message.asJSONString else { return .error(WalletLinkError.unableToSerializeMessageJSON) }

        return Internet.statusChanges
            .filter { $0.isOnline }
            .takeSingle()
            .flatMap { _ in self.connection.sendString(jsonString) }
            .flatMap { callback.subject.takeSingle() }
            .map { $0.type.isOK }
            .retry(3, delay: 1)
            .timeout(15, scheduler: ConcurrentDispatchQueueScheduler(qos: .userInitiated))
            .logError()
            .map { success in
                self.pendingCallbacks[callback.requestId] = nil
                return success
            }
            .catchError { err in
                self.pendingCallbacks[callback.requestId] = nil
                throw err
            }
    }

    // MARK: - Pending requests callback management

    private func createCallback() -> WalletLinkCallback {
        let requestId = callbackSequence.incrementAndGet()
        let subject = ReplaySubject<ClientResponseDTO>.create(bufferSize: 1)

        pendingCallbacks[requestId] = subject

        return WalletLinkCallback(requestId: requestId, subject: subject)
    }

    /// This is called when the host sends a response to a request initiated by the signer
    private func receivedClientResponse(_ response: ClientResponseDTO) {
        guard let requestId = response.id else { return assertionFailure("Invalid message response \(response)") }

        let subject = pendingCallbacks[requestId]

        pendingCallbacks[requestId] = nil

        subject?.onNext(response)
    }

    /// this is called when the host sends a request initiated by the host
    private func receivedServerRequest(_ request: ServerRequestDTO) {
        incomingRequestsSubject.onNext(request)
    }

    private func processIncomingData(_ incoming: WebIncomingDataType) {
        guard
            case let .text(jsonString) = incoming,
            let json = jsonString.jsonObject as? [String: Any],
            let typeString = json["type"] as? String,
            let type = ServerMessageType(rawValue: typeString)
        else {
            return assertionFailure("Unknown WalletLink type \(incoming)")
        }

        switch type {
        case .ok, .fail, .publishEventOK:
            guard let response = ClientResponseDTO.fromJSONString(jsonString) else {
                return assertionFailure("[walletlink] Invalid client response \(jsonString)")
            }

            receivedClientResponse(response)
        case .event:
            guard let request = ServerRequestDTO.fromJSONString(jsonString) else {
                return assertionFailure("[walletlink] Invalid server request \(jsonString)")
            }

            receivedServerRequest(request)
        }
    }
}
